app/models/manageiq/providers/redfish/physical_infra_manager/physical_server.rb

```diff
@@ -3,5 +3,10 @@ class PhysicalInfraManager::PhysicalServer < ::PhysicalServer
     def self.display_name(number = 1)
       n_('Physical Server (Redfish)', 'Physical Servers (Redfish)', number)
     end
+
+    def provider_object(connection = nil)
+      connection ||= ext_management_system.connect
```

Comment:
This line will leak a session if the `connection` argument is `nil`. Most Redfish systems have two properties that are relevant here:

 1. they usually support only a handful of concurrent sessions (for example, Dell's iDRAC supports only 8 concurrent sessions) and
 2. those sessions are long-lived (unused session is cleaned after about 30 minutes).

As a consequence, we can easily lock ourselves our for a significant amount of time if we do not terminate the sessions when we stop using them.

`ProviderObjectMixin` internals always call `provider_object` method with a non-`nil` connection (and that connection is properly terminated), so maybe we can simply drop this line? Or do we anticipate the method to be called with `connection == nil`?

Reply:
@tadeboro can you call `disconnect(connection)` if it wasn't passed in?  That should cover all of the cases

Reply:
Calling `disconnect` will plug the session leak, but it will also make the object returned form this method almost completely useless.

For example, let us say that we would like to update some BIOS settings on the server. Usually, we would call:

    server_model.with_provider_object do |server|
      server.patch(payload: { "Bios" => { "Setting" => "Value" } })
    end

One would think that this is the same as doing:

    server_model.provider_object.patch(payload: { "Bios" => { "Setting" => "Value" } })

But the second variant will fail with 401 error, because session has been terminated before we got our hands on the provider object. This means that we can only use the object returned from the `provider_object` for static field inspection, which severely limits the usefulness of having such object around.

I also checked the other providers and found out that OpenStack and oVirt provider do not handle `nil`s in their `provider_object` methods, so we would not be the first ones to require connection here. And oVirt is quite similar to our provider in this respect, since this is the only provider besides us that overrides ems' `disconnect` method.

Reply:
Ah okay gotcha, I'm good with not handling a nil connection then we should be pushing people towards using the block anyway

Reply:
Good point, thanks @tadeboro for noticing. I've ended up raising NotImplemented exception when connection is `nil` to make it more obvious it's erroring on purpose. Was struggling a bit with this when implementing because technically we could also implement `provider_object_release` and instruct user she needs to invoke it herself if called with nil. But I like your suggestion better because it wasn't really meant to be called without connection anyway. 

I think you just preserved us from a shitload of debugging in the future :)